[WIP] Add MM Grounding DINO

rziga opened the pull request:
# What does this PR do?

Fixes #37744.

It adds support for MM Grounding DINO and LLMDet (inference only).

* I've added a weight conversion script and a modular transformers implementation.
* The forward pass works with 1e-3 tolerance compared to the original (checked tiny and large checkpoints).
* The tests are copied and modified from Grounding DINO, but two integration tests fail currently (batched consistency and CPU-GPU consistency).

# TODO
[ ] fix failing tests

Tagging @qubvel

rziga committed:
first commit

Added modular implementation for MM Grounding DINO from starting point created by add-new-model-like. Added conversion script from mmdetection to huggingface.

TODO: Some tests are failing so that needs to be fixed.

rziga committed:
fixed a bug with modular definition of MMGroundingDinoForObjectDetection where box and class heads were not correctly assigned to inner model

rziga committed:
cleaned up a hack in the conversion script

rziga committed:
Fixed the expected values in integration tests

Cross att masking and cpu-gpu consistency tests are still failing however.

rziga committed:
changes for make style and quality

rziga committed:
add documentation

rziga committed:
clean up contrastive embedding

rziga committed:
add mm grounding dino to loss mapping

rziga committed:
add model link to config docstring

rziga committed:
hack fix for mm grounding dino consistency tests

rziga committed:
add special cases for unused config attr check

yonigozlan commented:
Hi @rziga ! Feel free to ping me when the PR is ready for review!

rziga commented:
Sorry, I got busy with life stuff for the last couple of weeks, so I didn't have time to finish it up. 

The model works, but I have to upload all the checkpoints to the hub and update the model doc before I mark it ready for review.

rziga committed:
add all models and update docs

rziga commented:
Pinging @yonigozlan and @qubvel as requested.

rziga committed:
update model doc to the new style

qubvel commented on docs/source/en/model_doc/mm-grounding-dino.md:
```diff
@@ -0,0 +1,156 @@
+
+from transformers import pipeline
+
+
+# Prepare pipeline
+model_id = "rziga/mm_grounding_dino_tiny_o365v1_goldg_v3det"
+pipe = pipeline("zero-shot-object-detection", model=model_id)
+
+# Run inference
+outputs = pipe(
+    "http://images.cocodataset.org/val2017/000000039769.jpg",
+    "a cat. a remote control.",
+    device="cuda" if torch.cuda.is_available() else "cpu",
+)
```
```suggestion
device = "cuda" if torch.cuda.is_available() else "cpu"

# Prepare pipeline
model_id = "rziga/mm_grounding_dino_tiny_o365v1_goldg_v3det"
pipe = pipeline("zero-shot-object-detection", model=model_id)

# Run inference
outputs = pipe(
    "http://images.cocodataset.org/val2017/000000039769.jpg",
    ["a cat", "a remote control"],
    device=device,
)
```

qubvel committed:
Use super_kwargs for modular config

qubvel committed:
Move init to the _init_weights function

qubvel committed:
Add copied from for tests

qubvel committed:
fixup

qubvel committed:
Merge branch 'main' into pr/rziga/37925

qubvel committed:
update typehints

qubvel committed:
Fix-copies for tests

qubvel committed:
fix-copies

qubvel committed:
Merge branch 'main' into add_mm_grounding_dino

qubvel committed:
Fix init test

qubvel committed:
fix snippets in docs

qubvel reviewed (commented):
Hello @rziga! Huge thanks for working on the model addition. It's really nice to see that the modular difference is minimal!

I pushed a few minor fixes:
  - Added super_kwargs for the config to minimize the modular difference even more.
  - Included `# Copied from` statements for the tests.
  - Fixed code snippets in the documentation.
  
The only things left are 
 - refactor the conversion script to make it similar to the library standards
 - prepare model cards on the HF Hub
 - make sure integration tests are pass on CI
 
Other than that, I believe the model is ready. Let me know if you have any questions and huge, huge sorry for the delay on our side 🤗

Lelege0 commented:
When is m_grounding_dino expected to be officially integrated into transformers and released? @qubvel  qubvel

qubvel commented:
Hi @Lelege0, I can't tell you the exact date, it depends on @rziga's bandwidth. It's at the top of my priority list at the moment, so I would be happy to merge it as soon as it's ready

qubvel committed:
fix consistency

qubvel committed:
fix consistency

Lelege0 commented:
> Hi @Lelege0, I can't tell you the exact date, it depends on @rziga's bandwidth. It's at the top of my priority list at the moment, so I would be happy to merge it as soon as it's ready

Thank you very much for your contribution!

rziga committed:
update conversion script

qubvel commented:
@rziga you can make the first one manually, and as soon as it's ready, you can update the rest with a simple script using `huggingface_hub`. Here is an example of how to upload just the README.md file

```
from huggingface_hub import HfApi
api = HfApi()

api.upload_file(
    path_or_fileobj="/path/to/local/folder/README.md",
    path_in_repo="README.md",
    repo_id="rziga/mm_grounding_dino_...",
    repo_type="model",
)
```

qubvel reviewed (approved):
Thanks for updating the conversion script! Everything looks good now 🚀 
Let's copy model cards to other models as well.

qubvel commented:
cc @ArthurZucker for final review

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_37925). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qubvel reviewed (approved):
Just a few more nit comments. 

By the way, @rziga, we usually move checkpoints to the organization that trained the model or to a `*-community` organization created by our team to maintain the checkpoints. Would it be ok for you after we get the final PR approval from @ArthurZucker?

In this particular case `openmmlab` seems to be not active on the Hub, so I can create `openmmlab-community` org and give you `write` permissions so you can move all the checkpoints and collections there (but only in case it's fine for you) 🤗

rziga committed:
fix nits in readme and remove old comments from conversion script

rziga commented:
I have no problem moving the checkpoints. I have no affiliation with any of the original model authors.

Also the LLMDet checkpoints, should probably be in a separate org. I don't know if there already exist an iSEE org, I can't find it. Perhaps @fushh (#37334) can tell where LLMDet checkpoints should be.

fushh commented:
@rziga Thanks for your valuable contributions🙏🙏🙏. Have you modified my models?
model: https://huggingface.co/fushh7/llmdet_swin_tiny_hf
model: https://huggingface.co/fushh7/llmdet_swin_base_hf
model: https://huggingface.co/fushh7/llmdet_swin_large_hf
If you modified them, please move your version to this collection. I will give you the write permissions. Otherwise, I will use my original models.
https://huggingface.co/collections/iSEE-Laboratory/llmdet-688475906dc235d5f1dc678e

rziga commented:
Hi @fushh . I have modified them. My conversion script copies parameters from mm-detection checkpoint directly and there are some changes in the contrastive bias parameters. 

In the original mmdet checkpoint (`https://huggingface.co/fushh7/LLMDet/resolve/main/tiny.pth`) each layer in class head has a different bias:
```py
import torch

original = torch.hub.load_state_dict_from_url(
    "https://huggingface.co/fushh7/LLMDet/resolve/main/tiny.pth?download=true",
    map_location="cpu",
)
for k in original["state_dict"]:
    if "bbox_head.cls_branches" in k:
        print(k, original["state_dict"][k])

>>>bbox_head.cls_branches.0.bias tensor([-4.5637])
>>>bbox_head.cls_branches.1.bias tensor([-4.5809])
>>>bbox_head.cls_branches.2.bias tensor([-4.5068])
>>>bbox_head.cls_branches.3.bias tensor([-4.4998])
>>>bbox_head.cls_branches.4.bias tensor([-4.4989])
>>>bbox_head.cls_branches.5.bias tensor([-4.4106])
>>>bbox_head.cls_branches.6.bias tensor([-5.1244])
```

Your hf converted checkpoints seem to only include one decoder bias:
```py
import safetensors.torch as st

converted = st.load_file("/home/user/.cache/huggingface/hub/models--fushh7--llmdet_swin_tiny_hf/snapshots/6719e6ec2b2f7f0f3fed035fdca7b1856a1107c9/model.safetensors", device="cpu")
for k in converted:
    if "class_embed" in k:
        print(k, converted[k])

>>>class_embed.0.bias tensor([-4.4106])
>>>model.encoder_output_class_embed.bias tensor([-5.1244])
```

Looking at your huggingface inference code, you don't seem to do a deep copy of your class embedding layer. In original GroundingDino this makes no difference, but here, each layer has its own bias parameter, so I think the bug comes from [here](https://github.com/iSEE-Laboratory/LLMDet/blob/67f0a6311d65077bd6fe843cce1cb327d5c9e8b8/hf_model/modeling_grounding_dino.py#L3003C9-L3003C95). 

Besides that, the config also requires changes to select the correct autoclass, so I think it's best to just move my checkpoints to you.

rziga commented:
@fushh I moved the models to your collection here: https://huggingface.co/collections/iSEE-Laboratory/llmdet-688475906dc235d5f1dc678e. 

You can remove me from the org now.

qubvel commented:
Hi @rziga, I created org for MM Grounding DINO checkpoints and invited you
 - https://huggingface.co/openmmlab-community

ArthurZucker commented on src/transformers/models/mm_grounding_dino/modular_mm_grounding_dino.py:
```diff
@@ -0,0 +1,258 @@
+import math
```
missing licence

ArthurZucker commented on src/transformers/models/mm_grounding_dino/modular_mm_grounding_dino.py:
```diff
@@ -0,0 +1,258 @@
+
+        self.model = MMGroundingDinoModel(config)
+
+        if config.decoder_cls_embed_share:
+            _class_embed = MMGroundingDinoContrastiveEmbedding(config)
+            self.class_embed = nn.ModuleList([_class_embed for _ in range(config.decoder_layers)])
+        else:
+            module_list = []
+            for _ in range(config.decoder_layers):
+                _class_embed = MMGroundingDinoContrastiveEmbedding(config)
+                module_list.append(_class_embed)
+            self.class_embed = nn.ModuleList(module_list)
+
+        if config.decoder_bbox_embed_share:
+            _bbox_embed = MMGroundingDinoMLPPredictionHead(
+                input_dim=config.d_model, hidden_dim=config.d_model, output_dim=4, num_layers=3
+            )
+            self.bbox_embed = nn.ModuleList([_bbox_embed for _ in range(config.decoder_layers)])
+        else:
+            module_list = []
+            for _ in range(config.decoder_layers):
+                _bbox_embed = MMGroundingDinoMLPPredictionHead(
+                    input_dim=config.d_model, hidden_dim=config.d_model, output_dim=4, num_layers=3
+                )
+                module_list.append(_bbox_embed)
+            self.bbox_embed = nn.ModuleList(module_list)
```
in general we should not have so many code paths! Each case would lead to a different Embedding.... Are there a model for each case?

qubvel replied:
ahh, I see! Yes, let's copy and modify the whole config class then

ArthurZucker commented on src/transformers/models/glm4v/image_processing_glm4v_fast.py:
```diff
@@ -188,9 +188,6 @@ def preprocess(
         images: ImageInput,
         **kwargs: Unpack[Glm4vFastImageProcessorKwargs],
     ) -> BatchFeature:
-        """
-        Preprocess an image or batch of images.
-        """
```
yes, let's remove this unrelated change

ArthurZucker reviewed (approved):
Very nice! Thanks for making it so easy to review with modular!

rziga committed:
add license

rziga committed:
remove unused config args

qubvel commented on src/transformers/models/mm_grounding_dino/modular_mm_grounding_dino.py:
```diff
@@ -203,16 +319,83 @@ def _init_weights(self, module):
+        self.backbone = MMGroundingDinoConvModel(backbone, position_embeddings)
+
+        # Create input projection layers
+        if config.num_feature_levels > 1:
```
Same here, can we remove this if/else? I suppose `num_feature_levels` > 1 for all checkpoints

rziga committed:
remove unnecessary if/else in model init

qubvel committed:
Merge branch 'main' into add_mm_grounding_dino

rziga committed:
fix quality

qubvel committed:
Update references

qubvel committed:
Merge branch 'main' into pr/rziga/37925

qubvel committed:
fix test

qubvel committed:
fixup

qubvel commented:
run-slow: grounding_dino, mm_grounding_dino
